adr/2018-06-12_Choose_a_serverless_framework.md

```diff
@@ -0,0 +1,42 @@
+### Claudia
+Claudia is a deployment utility built specifically for NodeJS and AWS Lambda. Its main design goal is to trade code portability and platform agnosticism for simplicity of workflow and the streamlining of basic tasks. Claudia also aims to implement features in ways that would seem intuitive to those familiar with the Node ecosystem, for instance modeling API Gateway as a JavaScript web server and routing code errors to 500 by default. "Behaving as JavaScript developers expect out of the box" cuts down on time spent learning new conventions in order to use the utility, something that is also helped by the large amount of documentation and examples available. It is not a framework, which cuts both ways: on one hand there are fewer constraints on code structure and low overhead, but on the other it could lead to poor code organization or a lack of extensibility down the line. [Claudia is able to handle requests and responses containing binary data out of the box.](https://claudiajs.com/tutorials/binary-content.html)
+
+### Apex
+Apex's has robust support for testing, with support for discrete dev/prod environments, function hooks, and pre-defined testing CLI commands out of the box. Apex's main selling point, is it's support for languages that AWS Lambda does not support by default such as Golang. Documentation is not as robust as that for Claudia or Serverless, and it abstracts away far less lambda setup.
+
+### Shep
+Like Claudia, Shep is a simple deployment tool, basically abstracting away much of the AWS deployment CLI and AWS Lambda setup. It's simplicity comes at the cost of functionality and extensibility and it lacks much of the documentation of the aforementioned frameworks and utilities.
+
+## Decision
+We will use Claudia/Claudia API Builder as a development utility/API Framework in order to cut down on unnecessary initial bloat and learning curve steepness. Of the four options, Claudia is the closest to Chalice in terms of feature-set while operating the most natively to the Node ecosystem. While Serverless performs all the functions of Claudia and more, it also introduces boilerplate and bloat to support features that Tilegarden does not require. Additionally, compared to Serverless, Claudia requires less work to configure binary data responses, which is crucial as serving images is Tilegarden's core function. Apex and Shep, for their part, lack much of the robust documentation, workflow abstraction, and JavaScript native-ness of the other options.
+
+## Consequences
+The consequences of this decision will be:
+* Must switch frameworks if switching serverless provider
+* Potentially more tedious resource provisioning in the short term
```

Comment:
I'm not quite clear on what this means.  Specifically "tedious" and "in the short term".  Is the idea that we'll be doing provisioning via "follow these written steps involving AWS CLI commands" initially then later automate it with a provisioning framework (e.g. terraform)?  In that case I'd say there are two pieces--additional provisioning automation work (vs. a framework that included a configurable default setup for that) and more manual provisioning in the period before we do the automation work.
Which is to say, I would replace "tedious" with something more concrete (i.e. "manual") and be clearer about what "the short term" means.